plugins/outputs/otlp/client.go

```diff
@@ -0,0 +1,314 @@
+// dialing, this function uses its a new context and the same timeout
+// used for Store().
+func (c *Client) getConnection(ctx context.Context) (_ *grpc.ClientConn, retErr error) {
+	if c.conn != nil {
+		return c.conn, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	useAuth := c.url.Scheme != "http"
+	_ = level.Debug(c.logger).Log(
+		"msg", "new OTLP connection",
+		"auth", useAuth,
+		"url", c.url.String(),
+		"timeout", c.timeout)
+
+	dopts := []grpc.DialOption{
+		grpc.WithBlock(), // Wait for the connection to be established before using it.
+		grpc.WithDefaultServiceConfig(serviceConfig),
+	}
+	if useAuth {
+		var tcfg tls.Config
+		if len(c.rootCertificates) != 0 {
+			certPool := x509.NewCertPool()
+
+			for _, cert := range c.rootCertificates {
+				bs, err := ioutil.ReadFile(cert)
+				if err != nil {
+					return nil, fmt.Errorf("could not read certificate authority certificate: %s: %w", cert, err)
+				}
+
+				ok := certPool.AppendCertsFromPEM(bs)
+				if !ok {
+					return nil, fmt.Errorf("could not parse certificate authority certificate: %s: %w", cert, err)
+				}
+			}
+
+			tcfg = tls.Config{
+				ServerName: c.url.Hostname(),
+				RootCAs:    certPool,
+			}
+		}
+		_ = level.Debug(c.logger).Log(
+			"msg", "TLS configured",
+			"server", c.url.Hostname(),
+			"root_certs", fmt.Sprint(c.rootCertificates),
+		)
+		dopts = append(dopts, grpc.WithTransportCredentials(credentials.NewTLS(&tcfg)))
+	} else {
+		dopts = append(dopts, grpc.WithInsecure())
+	}
+	if c.compressor != "" && c.compressor != "none" {
+		dopts = append(dopts, grpc.WithDefaultCallOptions(grpc.UseCompressor(c.compressor)))
+	}
+	address := c.url.Hostname()
+	if len(c.url.Port()) > 0 {
+		address = net.JoinHostPort(address, c.url.Port())
+	}
+	conn, err := grpc.DialContext(ctx, address, dopts...)
+	c.conn = conn
+	if err != nil {
+		_ = level.Debug(c.logger).Log(
+			"msg", "connection status",
+			"address", address,
+			"err", err,
+		)
+		return nil, err
+	}
+
+	return conn, err
+}
+
+// Selftest sends an empty request the endpoint.
+func (c *Client) Selftest(ctx context.Context) error {
+	// Loop until the context is canceled, allowing for retryable failures.
+	for {
+		conn, err := c.getConnection(ctx)
+
+		if err == nil {
+			service := metricsService.NewMetricsServiceClient(conn)
+			empty := &metricsService.ExportMetricsServiceRequest{}
+
+			_, err = service.Export(c.grpcMetadata(ctx), empty)
+			if err == nil {
+				return nil
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if isRecoverable(err) {
+				_ = level.Info(c.logger).Log("msg", "selftest recoverable error, still trying", "err", err)
+				continue
+			}
+		}
+		return fmt.Errorf(
+			"non-recoverable failure in selftest: %s",
+			truncateErrorString(err),
+		)
+	}
+}
+
+// Store sends a batch of samples to the endpoint.
+func (c *Client) Store(req *metricsService.ExportMetricsServiceRequest) error {
```

Comment:
Please do not export this function as it should only be used within the boundaries of this package.

Reply:
Fixed.